### PR TITLE
[Purity] Collect all blobs first before marking them (and some debug logs)

### PR DESCRIPTION
**Pull request by fische**
This PR rewrites a bit the way we mark blobs in purity by first collecting them all and then marking them as referenced. This reduces a bit the complexity of the code and allows us to log the input and output blobs that have been found more easily.

### CODE REVIEW

**Comment by fische** [comment]
By the way, I've just run the new and old versions at the same tiem (with `dry_run`) and they are both doing approximately the same:
```
# old
$ ./purity -u localhost:7775 one --min_age 52h --log_file ~/purity.old.log --file_verbosity debug --replication_factor 2 --dry_run
...
17:23:44.999  NOTICE: Found 310101 live action results and 2416633 referenced blobs
17:23:44.999 WARNING: Missing inputs for 5813 action results
17:23:44.999  NOTICE: Determining blobs to remove...
17:23:47.713  NOTICE: Would delete 163956 blobs, total size 1.4 GB
17:23:47.713  NOTICE: Would remove 3 actions:
17:23:47.772  NOTICE: Found 30 underreplicated blobs, total size 23 kB
17:23:47.772  NOTICE: Found 19 underreplicated action results, total size 521 B
17:23:47.772  NOTICE: Complete in 17m58s!

# new
$ ./purity -u localhost:7775 one --min_age 52h --log_file ~/purity.new.log --file_verbosity debug --replication_factor 2 --dry_run
...
17:24:44.256  NOTICE: Found 310101 live action results and 2419810 referenced blobs
17:24:44.256 WARNING: Missing inputs for 5813 action results
17:24:44.256  NOTICE: Determining blobs to remove...
17:24:46.168  NOTICE: Would delete 162039 blobs, total size 1.4 GB
17:24:46.168  NOTICE: Would remove 3 actions:
17:24:46.205  NOTICE: Found 30 underreplicated blobs, total size 23 kB
17:24:46.205  NOTICE: Found 19 underreplicated action results, total size 521 B
17:24:46.205  NOTICE: Complete in 18m56s!
```

**Comment by fische** [comment]
Will add some documentation on this as soon as I have time, as it's a bit hard to read/understand at the moment

**Comment by fische** [comment]
This is again available for review

File: purity/gc/gc.go [hunk 1]
```diff
@@ -225,6 +220,13 @@ func min(a, b int) int {
 	return b
 }
 
+// MarkReferencedBlobs traverses the internal list of "live" action results
+// and for each of them marks all the blobs they point to (either directly or
+// indirectly) as "referenced" by adding their digest to an internal map. See
+// `RemoveBlobs` on how this map is then used.
+// Missing input blobs are non-fatal, but missing output blobs will get the
```
Comment (Garbett1) [inline purity/gc/gc.go:227]: >Missing input blobs are non-fatal

Given the nasty bucket clearing incident, should we re-consider this?

**Pull request merged**